File: 2545-height-of-binary-tree-after-subtree-removal-queries/2545-height-of-binary-tree-after-subtree-removal-queries.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def treeQueries(self, root: Optional[TreeNode], queries: List[int]) -> List[int]:
        prefix = Counter()
        suffix = Counter()

        def pre_order(node, height = 0):
            if not node:
                return
            
            prefix[node.val] = self.max_height
            self.max_height = max(self.max_height, height)

            pre_order(node.left, height+1)
            pre_order(node.right, height+1)
        
        def suf_order(node, height = 0):
            if not node:
                return
            
            suffix[node.val] = self.max_height
            self.max_height = max(self.max_height, height)

            suf_order(node.right, height+1)
            suf_order(node.left, height+1)
        
        self.max_height = 0
        pre_order(root)
        self.max_height = 0
        suf_order(root)

        return [max(prefix[val], suffix[val]) for val in queries]

File: 2545-height-of-binary-tree-after-subtree-removal-queries/2545-height-of-binary-tree-after-subtree-removal-queries.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Integer> preMap;
    Map<Integer, Integer> sufMap;
    int maxHeight;
    public int[] treeQueries(TreeNode root, int[] queries) {
        this.preMap = new HashMap<>();
        this.sufMap = new HashMap<>();
        this.maxHeight = 0;
        preSum(root, 0);
        this.maxHeight = 0;
        sufSum(root, 0);

        int[] res = new int[queries.length];
        int idx = 0;
        for (int q : queries) {
            res[idx++] = Math.max(preMap.get(q), sufMap.get(q));
        }

        return res;
    }

    void preSum(TreeNode root, int height) {
        if (root == null) {
            return;
        }

        preMap.put(root.val, maxHeight);
        maxHeight = Math.max(maxHeight, height);

        preSum(root.left, height + 1);
        preSum(root.right, height + 1);
    }

    void sufSum(TreeNode root, int height) {
        if (root == null) {
            return;
        }

        sufMap.put(root.val, maxHeight);
        maxHeight = Math.max(maxHeight, height);

        sufSum(root.right, height + 1);
        sufSum(root.left, height + 1);
    }
}